File: integration-tests/TestAPI-bindings/kotlin/src/test/kotlin/com/cricut/testapi/AsyncFunctionTests.kt
package com.cricut.testapi

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import kotlin.math.abs
import kotlinx.coroutines.*
import kotlinx.coroutines.test.*
import org.junit.jupiter.api.BeforeAll
import kotlin.coroutines.resume
import kotlin.time.Duration.Companion.seconds

internal class AsyncFunctionTests {
    companion object {
        @BeforeAll
        @JvmStatic
        fun beforeAll() {
            println(ProcessHandle.current().pid())
            println()
            // java.lang.Thread.sleep(15000)
        }
    }

    @Test
    fun testConst42() = runTest {
        assertEquals(42, AsyncFunctions.const42())
    }

    @Test
    fun testAbs() = runTest(timeout = 1000000.seconds) {
        val theAbs = AsyncFunctions.abs
        val abs3 = theAbs(-3)
        assertEquals(3, abs3)
    }

    @Test
    fun testIntCompose() = runTest(timeout = 100000.seconds) {
        val intCompose = AsyncFunctions.intCompose
        val composed = intCompose({ it + 1 }, { it * 3 })
        assertEquals(10, composed(3))
        assertEquals(7, composed(2))
    }

    object ComposeError : Error()

    @Test
    fun testExceptionInCompose() = runTest(timeout = 100000.seconds) {
        val composed = AsyncFunctions.intCompose({ throw ComposeError }, { it })
        assertThrows<ComposeError> { composed(3) }
    }

    @Test
    fun testTheRestOfTheFunctions() = runTest {
        assertEquals(8.5, AsyncFunctions.add3Things(3f, 4.5, 1))
        assertEquals(listOf("a", "b", "c", "d"), AsyncFunctions.makeList("a", "b", "c", "d"))
        assertEquals(84, AsyncFunctions.fifthThing("hi", 1, 1.0, "...") { 84 }())
        assertEquals(17, AsyncFunctions.sixthThing("hi", 1, 1.0, "...", { 84 }, 17))
    }

    @Test
    fun testPassingFunctionsToSwift() = runTest(timeout = 100000.seconds) {
        assertEquals("8", AsyncFunctions.exercise0 { 8 })
        assertEquals("3", AsyncFunctions.exercise1 { abs(it) })
        assertEquals("25", AsyncFunctions.exercise2 { f, g -> { f(g(it)) } })
        assertEquals("7.4", AsyncFunctions.exercise3 { a, b, c -> a + b + c })
        assertEquals("""["a", "b", "c", "d"]""", AsyncFunctions.exercise4 { a, b, c, d -> listOf(a, b, c, d) })
        assertEquals("83", AsyncFunctions.exercise5 { _, _, _, _, f -> f })
        assertEquals("42", AsyncFunctions.exercise6 { _, _, _, _, _, i -> i })
    }

    @Test
    fun testSwiftThrows(): Unit = runTest {
        val result = runCatching {
            AsyncFunctions.willThrow()
        }
        assertInstanceOf(Error::class.java, result.exceptionOrNull())
        assertEquals("TheError()", result.exceptionOrNull()?.message)
    }
}

File: integration-tests/TestAPI-bindings/c-sharp/Cricut.TestAPI.Tests/AsyncFunctionTests.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Xunit;

namespace Cricut.TestAPI.Tests {
    public class AsyncFunctionTests {
        // Helper to avoid warnings about trivial async functions
        private static async Task<T> Async<T>(T result) {
            await Task.Yield();
            return result;
        }

        [Fact]
        async Task testAsyncConst42() {
            Assert.Equal(42, await AsyncFunctions.Const42());
        }

        [Fact]
        async Task testAsyncAbs() {
            Assert.Equal(3, await AsyncFunctions.Abs(-3));
        }

        [Fact]
        async Task testAsyncIntCompose() {
            var composed = AsyncFunctions.IntCompose(async (x) => await Async(x + 1), async (x) => await Async(x * 3));
            Assert.Equal(10, await composed(3));
            Assert.Equal(7, await composed(2));
        }

        class ComposeError: Exception {}

        [Fact]
        async Task testAsyncExceptionInCompose() {
            var composed = AsyncFunctions.IntCompose(_=> throw new ComposeError(), async (x) => await Async(x));
            await Assert.ThrowsAsync<ComposeError>(() => composed(3));
        }

        [Fact]
        async Task testAsyncTheRestOfTheFunctions() {
            Assert.Equal(8.5, await AsyncFunctions.Add3Things(3f, 4.5, 1));
            Assert.Equal(new string[] { "a", "b", "c", "d" }, await AsyncFunctions.MakeList("a", "b", "c", "d"));
            Assert.Equal(84, await (await AsyncFunctions.FifthThing("hi", 1, 1.0, "...", async () => await Async(84)))());
            Assert.Equal(17, await AsyncFunctions.SixthThing("hi", 1, 1.0, "...", async () => await Async(84), 17));
        }

        [Fact]
        async Task testAsyncPassingFunctionsToSwift() {
            Assert.Equal("8", await AsyncFunctions.Exercise0(async () => await Async(8)));
            Assert.Equal("3", await AsyncFunctions.Exercise1(async x => await Async(-x)));
            Assert.Equal("25", await AsyncFunctions.Exercise2((f, g) => async x => await f(await g(x))));
            Assert.Equal("7.4", await AsyncFunctions.Exercise3(async (a, b, c) => await Async(a + b + c)));
            Assert.Equal("[\"a\", \"b\", \"c\", \"d\"]", await AsyncFunctions.Exercise4(async (a, b, c, d) => await Async(new[] { a, b, c, d })));
            Assert.Equal("83", await AsyncFunctions.Exercise5(async (_, _, _, _, f) => await Async(f)));
            Assert.Equal("42", await AsyncFunctions.Exercise6(async (_, _, _, _, _, i) => await Async(i)));
        }

        [Fact]
        async Task testAsyncSwiftThrows() {
            await Assert.ThrowsAsync<Exception>(async () => await AsyncFunctions.WillThrow());
        }

        [Fact]
        async Task TestThunkMaker() {
            var timesCalled = 0;
            var thunk = AsyncFunctions.ThunkTwiceMaker(async () => { await Task.Yield(); timesCalled++; });
            await thunk();
            Assert.Equal(2, timesCalled);
        }
    }
}
